filclient/filclient.go

```diff
@@ -853,221 +857,113 @@ func (fc *FilClient) RetrieveContent(ctx context.Context, miner address.Address,
 	}
-
-	// Allocate a lane on our payment channel, usually you want a new lane per retrieval
-	lane, err := fc.pchmgr.AllocateLane(pchaddr)
+	pchLane, err := fc.pchmgr.AllocateLane(pchAddr)
 	if err != nil {
 		return nil, xerrors.Errorf("failed to allocate lane: %w", err)
 	}
 
-	// Use the data transfer protocol to propose the retrieval deal
-	sel := shared.AllSelector()
-	var vouch datatransfer.Voucher = proposal
-	chanid, err := fc.dataTransfer.OpenPullDataChannel(ctx, mpid, vouch, proposal.PayloadCID, sel)
+	// Submit the retrieval deal proposal to the miner
+	chanid, err := fc.dataTransfer.OpenPullDataChannel(ctx, mpID, proposal, proposal.PayloadCID, shared.AllSelector())
 	if err != nil {
 		return nil, err
 	}
 
-	unsub := fc.dataTransfer.SubscribeToEvents(func(event datatransfer.Event, channelState datatransfer.ChannelState) {
-		if channelState.ChannelID() == chanid {
-			fmt.Printf("[%s] event: %s %d %s %s\n", time.Now().Format("15:04:05"), event.Message, event.Code, datatransfer.Events[event.Code], event.Timestamp.Format("15:04:05"))
-			fmt.Printf("channelstate: %s %s\n", datatransfer.Statuses[channelState.Status()], channelState.Message())
+	// Used to limit prints to the console about received status
+	lastReceivedUpdate := time.Now()
+
+	// The next nonce (incrementing unique ID starting from 0) for the next voucher
+	var nonce uint64 = 0
+
+	// Set up incoming events handler
+
+	// dtRes receives either an error (failure) or nil (success) which is waited
+	// on and handled below before exiting the function
+	dtRes := make(chan error)
+	unsubscribe := fc.dataTransfer.SubscribeToEvents(func(event datatransfer.Event, state datatransfer.ChannelState) {
+		switch event.Code {
+		case datatransfer.NewVoucherResult:
+			switch resType := state.LastVoucherResult().(type) {
+			case *retrievalmarket.DealResponse:
+				switch resType.Status {
+				case retrievalmarket.DealStatusAccepted:
+					log.Info("deal accepted")
+
+				// Respond with a payment voucher when funds are requested
+				case retrievalmarket.DealStatusFundsNeeded:
+					log.Infof("sending payment voucher: %v", resType.PaymentOwed)
+
+					totalPayment = types.BigAdd(totalPayment, resType.PaymentOwed)
+
+					vres, err := fc.pchmgr.CreateVoucher(ctx, pchAddr, paych.SignedVoucher{
+						ChannelAddr: pchAddr,
+						Lane:        pchLane,
+						Nonce:       nonce,
+						Amount:      totalPayment,
+					})
+					if err != nil {
+						dtRes <- err
+					}
+
+					if types.BigCmp(vres.Shortfall, big.NewInt(0)) > 0 {
+						dtRes <- xerrors.Errorf("not enough funds remaining in payment channel (shortfall = %s)", vres.Shortfall)
+					}
+
+					if err := fc.dataTransfer.SendVoucher(ctx, chanid, &retrievalmarket.DealPayment{
+						ID:             proposal.ID,
+						PaymentChannel: pchAddr,
+						PaymentVoucher: vres.Voucher,
+					}); err != nil {
+						dtRes <- xerrors.Errorf("failed to send payment voucher: %w", err)
+					}
+
+					nonce++
```

Comment:
Are calls to this event handler potentially happening in parallel to eachother? Its probably not an issue in practice, but we could have data consistency issues if we ran this event code twice in parallel

Reply:
You mean pushing to dtRes? Or the entire callback in SubscribeToEvents?

Reply:
the entire callback really, im just looking at this nonce incrementing and thinking that we could end up with things out of order, or having the wrong values if we (somehow) end up doing these in parallel.

Reply:
is this still a problem? i'm not sure quite exactly what you're getting at. the whole callback is guarded by a mutex now, does that help?